### ISSUE TITLE
ValueError when slicing with empty masked arrays 

### ISSUE THREAD

**Issue by jrueb** [scikit-hep/awkward#1288]
### Version of Awkward Array

1.8.0rc3

### Description and code to reproduce

Please consider the following:
```python
import awkward as ak
array = ak.Array([[5.,6.],[7.]])
masked = array.mask[[[True, True], [True]]]
empty = masked[[False, False]]
```
Now the following works:
```python
array[ak.argsort(array)]
masked[ak.argsort(masked)]
```
But this doesn't (giving `ValueError: only arrays of integers or booleans may be used as a slice`)
```python
empty[ak.argsort(empty)]
```

I see two inconsistencies here:
- `ak.argsort(masked)` returns an array of values type int, while `ak.argsort(empty)` returns an array of values type float. In both cases it should be int.
- Slicing an array that doesn't have a mask with an array of values type float works, but doing the same with a masked array raises an error.

**Comment by agoose77** [comment]
Hi @jrueb, thanks for the report!

Could you elaborate on your last point? Namely:
> Slicing an array that doesn't have a mask with an array of values type float works, but doing the same with a masked array raises an error.

I assume you've seen this during your debugging, but if I take your statement to the simplest case of a floating point index array, then I can't reproduce it.

RE `ak.argsort(empty)`, this is an artefact of returning `shallow_copy` where the length is 0:  

https://github.com/scikit-hep/awkward-1.0/blob/aaded1ccee9bbe24048d084a2141fe410a576ff7/src/libawkward/array/ListOffsetArray.cpp#L1920

I assume that we do this to preserve the dimensionality of the result in this case. However, we should not be returning the array with the original dtype.

@jpivarski this is a wider policy issue that I'm not clear on. We have the following behaviour for `len(this) == 0` in v1 and v2, for these layouts:
- `ak.layout.ListOffsetArray` 
  ```python
  <ByteMaskedArray valid_when="true">
    <mask><Index8 i="[]" offset="0" length="0" at="0x55a45347bca0"/></mask>
    <content><NumpyArray format="d" shape="0" data="" at="0x55a45335fe60"/></content>
  </ByteMaskedArray>
  ```
- `ak.layout.RegularArray` - `<NumpyArray format="l" shape="0" data="" at="0x000000000000"/>`
- `ak._v2.contents.ListOffsetArray` - `<NumpyArray dtype='int64' len='0'>[]</NumpyArray>`
- `ak._v2.contents.RegularArray` - `<NumpyArray dtype='int64' len='0'>[]</NumpyArray>`

So, currently only `ak.layout.ListOffsetArray` tries to preserve the structure of the array. Personally I think this makes the most sense - I would intuitively expect the `argsort` result to have the same number of dimensions as the array upon which it operates. This would mean changing v2, and v1 `ak.layout.RegularArray`. Do you agree here?

**Comment by jrueb** [comment]
> I assume you've seen this during your debugging, but if I take your statement to the simplest case of a floating point index array, then I can't reproduce it.

I just noticed it's a bit more specific than I thought. What I saw was that
```python
empty2 = array[[False, False]]
empty2[ak.argsort(empty2)]
```
works. `empty2` has value type `float` and so does `ak.argsort(empty2)`. From the error message "only arrays of integers or booleans may be used as a slice" I would expect this not to work. However, looking at this now, it only happens with one dimensional empty array slices. So this might also be meant as a corner case, where it's fine not to raise an error?

**Comment by jpivarski** [comment]
> I would intuitively expect the `argsort` result to have the same number of dimensions as the array upon which it operates. This would mean changing v2, and v1 `ak.layout.RegularArray`. Do you agree here?

@agoose77, you're right: `ak.argsort` (and `ak._v2.argsort`) should return an array with the same dimensions as the input, even if it's empty. Especially now that we're requiring the output Form of every operation to be independent of any buffer values or shapes, so that a Dask DAG can be examined independently of any concrete arrays. Short-circuits like

https://github.com/scikit-hep/awkward-1.0/blob/aaded1ccee9bbe24048d084a2141fe410a576ff7/src/libawkward/array/ListOffsetArray.cpp#L1912-L1921

or even

https://github.com/scikit-hep/awkward-1.0/blob/5e49baf2f20fb52a66aaa121d521932f5980cd27/src/awkward/_v2/contents/listoffsetarray.py#L1030-L1045

might result in different Forms for length 0 arrays than any other length, and when building a Dask DAG, we don't know if the length will be 0. (Even if `self._offsets.length` returns `UnknownLength`, [any comparison operator on this returns false](https://github.com/scikit-hep/awkward-1.0/blob/5e49baf2f20fb52a66aaa121d521932f5980cd27/src/awkward/_v2/_typetracer.py#L28-L81), so the DAG-building step wouldn't enter into the `if` statement, but the real array might.)

**Comment by jpivarski** [comment]
> `empty2` has value type `float` and so does `ak.argsort(empty2)`. From the error message "only arrays of integers or booleans may be used as a slice" I would expect this not to work.

@jrueb, what might be confusing is the distinction between an array with zero length and an [EmptyArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.EmptyArray.html), which has zero length and unknown type.

```python
>>> array = ak.Array(["This", "is", "not", "empty", "or", "unknown", "type."])
>>> array[[]]
<Array [] type='0 * string'>

>>> ak.Array([])
<Array [] type='0 * unknown'>
```

If you get an array of type `0 * float64` by slicing, you shouldn't be able to use that as a slicer because it has type floating point—it's unclear how you intend for it to work as a slicer.

If you have an array of type `0 * unknown`, the type it gets can be determined from how it's used.

```python
>>> ak.Array(["To", "be", "sliced."])[unknown_empty]
<Array [] type='0 * string'>
```

That might be why you got a different result in two cases. Be sure to check the `.type`!

**Comment by agoose77** [comment]
@jpivarski I think this is the issue I was thinking of!

**Comment by jpivarski** [comment]
Thanks, @agoose77, this _is_ the same issue as #1386, and it's fixed by the same PR #1387.

I just verified that the initial problem is gone:

```python
>>> import awkward as ak
>>> array = ak.Array([[5.,6.],[7.]])
>>> masked = array.mask[[[True, True], [True]]]
>>> empty = masked[[False, False]]
>>> array[ak.argsort(array)]
<Array [[5, 6], [7]] type='2 * var * float64'>
>>> masked[ak.argsort(masked)]
<Array [[5, 6], [7]] type='2 * var * ?float64'>
>>> empty[ak.argsort(empty)]
<Array [] type='0 * option[var * ?float64]'>
```

(Also for v2.)

So I'll mark this as one of the issues that will get closed by the PR.
